// @near-js/client/lib/commonjs/transactions/actions.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deployContract = exports.deleteAccount = exports.deleteAccessKey = exports.addFunctionCallAccessKey = exports.addFullAccessKey = exports.stake = exports.transfer = exports.functionCall = void 0;
const composers_1 = require("./composers/index.cjs");
/**
 * Make a function call against a contract
 * @param sender transaction signer
 * @param receiver target account/contract
 * @param method method to be invoked
 * @param args method arguments
 * @param gas attached gas
 * @param deposit attached deposit in yN
 * @param blockReference block ID/finality
 * @param deps sign-and-send dependencies
 */
function functionCall({ sender, receiver, method, args, gas, deposit, blockReference, deps }) {
    return composers_1.SignedTransactionComposer.init({ sender, receiver, deps })
        .functionCall(method, args, gas, deposit)
        .signAndSend(blockReference);
}
exports.functionCall = functionCall;
/**
 * Send Near from one account to another
 * @param sender account sending Near
 * @param receiver account receiving Near
 * @param amount Near to send in yN
 * @param blockReference block ID/finality
 * @param deps sign-and-send dependencies
 */
function transfer({ sender, receiver, amount, blockReference, deps }) {
    return composers_1.SignedTransactionComposer.init({ sender, receiver, deps })
        .transfer(amount)
        .signAndSend(blockReference);
}
exports.transfer = transfer;
/**
 * Stake Near with the specified validator
 * @param account account staking Near
 * @param amount Near to stake in yN
 * @param publicKey public key for the target validator
 * @param blockReference block ID/finality
 * @param deps sign-and-send dependencies
 */
function stake({ account, amount, publicKey, blockReference, deps }) {
    return composers_1.SignedTransactionComposer.init({ sender: account, receiver: account, deps })
        .stake(amount, publicKey)
        .signAndSend(blockReference);
}
exports.stake = stake;
/**
 * Add a full access key to an account
 * @param account account to which the FAK is added
 * @param publicKey public key string for the new FAK
 * @param blockReference block ID/finality
 * @param deps sign-and-send dependencies
 */
function addFullAccessKey({ account, publicKey, blockReference, deps }) {
    return composers_1.SignedTransactionComposer.init({ sender: account, receiver: account, deps })
        .addFullAccessKey(publicKey)
        .signAndSend(blockReference);
}
exports.addFullAccessKey = addFullAccessKey;
/**
 * Add a function call access key to an account
 * @param account account to which the access key is added
 * @param publicKey public key string for the new access key
 * @param contract contract on which methods may be invoked
 * @param methodNames set of methods which may be invoked
 * @param allowance maximum amount of Near which can be attached to a transaction signed with this key
 * @param blockReference block ID/finality
 * @param deps sign-and-send dependencies
 */
function addFunctionCallAccessKey({ account, publicKey, contract, methodNames, allowance, blockReference, deps }) {
    return composers_1.SignedTransactionComposer.init({ sender: account, receiver: account, deps })
        .addFunctionCallAccessKey(publicKey, contract, methodNames, allowance)
        .signAndSend(blockReference);
}
exports.addFunctionCallAccessKey = addFunctionCallAccessKey;
/**
 * Remove the specified access key from an account
 * @param account account from which the access key will be removed
 * @param publicKey public key string of the access key to be removed
 * @param blockReference block ID/finality
 * @param deps sign-and-send dependencies
 */
function deleteAccessKey({ account, publicKey, blockReference, deps }) {
    return composers_1.SignedTransactionComposer.init({ sender: account, receiver: account, deps })
        .deleteKey(publicKey)
        .signAndSend(blockReference);
}
exports.deleteAccessKey = deleteAccessKey;
/**
 * Delete an account; account funds will be transferred to the designated beneficiary
 * @param account account from which the access key will be removed
 * @param publicKey public key string of the access key to be removed
 * @param blockReference block ID/finality
 * @param deps sign-and-send dependencies
 */
function deleteAccount({ account, beneficiaryId, blockReference, deps }) {
    return composers_1.SignedTransactionComposer.init({ sender: account, receiver: account, deps })
        .deleteAccount(beneficiaryId)
        .signAndSend(blockReference);
}
exports.deleteAccount = deleteAccount;
/**
 * Deploy contract code to an account
 * @param account account to which the contract code will be deployed
 * @param code WASM code as byte array
 * @param blockReference block ID/finality
 * @param deps sign-and-send dependencies
 */
function deployContract({ account, code, blockReference, deps }) {
    return composers_1.SignedTransactionComposer.init({ sender: account, receiver: account, deps })
        .deployContract(code)
        .signAndSend(blockReference);
}
exports.deployContract = deployContract;
