// @near-js/providers/lib/commonjs/fetch_json.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchJsonRpc = void 0;
const types_1 = require("@near-js/types");
const exponential_backoff_1 = require("exponential-backoff");
const isomorphic_unfetch_1 = __importDefault(require("isomorphic-unfetch"));
const BACKOFF_MULTIPLIER = 1.5;
const RETRY_NUMBER = 10;
const retryConfig = {
    numOfAttempts: RETRY_NUMBER,
    timeMultiple: BACKOFF_MULTIPLIER,
    retry: (e) => {
        if ([503, 408].includes(e.cause)) {
            return true;
        }
        if (e.toString().includes('FetchError') || e.toString().includes('Failed to fetch')) {
            return true;
        }
        return false;
    }
};
class ProviderError extends Error {
    cause;
    constructor(message, options) {
        super(message, options);
    }
}
/**
 * Performs an HTTP request to an RPC endpoint
 * @param url URL for the HTTP request
 * @param json Request body
 * @param headers HTTP headers to include with the request
 * @returns Promise<any> }arsed JSON response from the HTTP request.
 */
async function fetchJsonRpc(url, json, headers) {
    const response = await (0, exponential_backoff_1.backOff)(async () => {
        const res = await (0, isomorphic_unfetch_1.default)(url, {
            method: 'POST',
            body: JSON.stringify(json),
            headers: { ...headers, 'Content-Type': 'application/json' }
        });
        const { ok, status } = res;
        if (!ok) {
            throw new ProviderError(await res.text(), { cause: status });
        }
        if (status === 503) {
            throw new ProviderError(`${url} unavailable`, { cause: status });
        }
        else if (status === 408) {
            throw new ProviderError('Unused connection', { cause: status });
        }
        return res;
    }, retryConfig);
    if (!response) {
        throw new types_1.TypedError(`Exceeded ${RETRY_NUMBER} attempts for ${url}.`, 'RetriesExceeded');
    }
    return await response.json();
}
exports.fetchJsonRpc = fetchJsonRpc;
