// @near-js/client/lib/commonjs/funded_account.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFundedTestnetAccount = void 0;
const isomorphic_fetch_1 = __importDefault(require("isomorphic-fetch"));
const constants_1 = require("./constants.cjs");
/**
 * Create a new funded testnet account via faucet REST endpoint
 *  (e.g. create `new.testnet` with a preset amount of Near)
 * @param endpointUrl REST endpoint for the funded testnet account creation (defaults to the current Near Contract Helper endpoint)
 * @param newAccount name of the created account
 * @param newPublicKey public key for the created account's initial full access key
 */
async function createFundedTestnetAccount({ newAccount, newPublicKey, endpointUrl = constants_1.KITWALLET_FUNDED_TESTNET_ACCOUNT_ENDPOINT, }) {
    const res = await (0, isomorphic_fetch_1.default)(endpointUrl, {
        method: 'POST',
        body: JSON.stringify({
            newAccountId: newAccount,
            newAccountPublicKey: newPublicKey,
        }),
        headers: { 'Content-Type': 'application/json' },
    });
    const { ok, status } = res;
    if (!ok) {
        throw new Error(`Failed to create account on ${endpointUrl}: ${status}`);
    }
    return await res.json();
}
exports.createFundedTestnetAccount = createFundedTestnetAccount;
